spec/core/ics-033-multi-hop/README.md

```diff
@@ -0,0 +1,571 @@
+        Value: connectionEndBytes,
+        Proof: connectionEndProof,
+    }
+}
+
+// queryMinimumConsensusHeight returns the minimum height within the provided
+// range at which a valid consensusState exists (processedHeight) and the
+// corresponding consensus state height (consensusHeight).
+func queryMinimumConsensusHeight(
+    chain Chain,
+    minConsensusHeight exported.Height,
+    limit uint64,
+) (
+    processedHeight exported.Height,
+    consensusHeight exported.Height
+) {
+
+    // find the minimum height consensus state
+    consensusHeight := minConsensusHeight
+    for i := uint64(0); i < limit; i++ {
+        key := host.FullClientKey(clientID, ibctm.ProcessedHeightKey(consensusHeight))
+        consensusStateHeightBytes, _ := abortTransactionUnless(chain.QueryStateAtHeight(key, chain.LastHeader.Header.Height, false))
+
+        if consensusStateHeightBytes != nil {
+            proofHeight := abortTransactionUnless(clienttypes.ParseHeight(string(consensusStateHeightBytes)))
+            return proofHeight, consensusHeight
+        }
+        consensusHeight = consensusHeight.Increment()
+    }
+
+    return nil, nil
+}
+```
+
+### Multi-hop Proof Verification Steps
+
+The following outlines the general proof verification steps specific to a multi-hop IBC message.
+
+1. Unpack the multihop proof bytes into consensus states, connection states and channel/commitment proof data.
+2. Check the counterparty client on the receiving end is active and the client height is greater than or equal to the proof height.
+3. Iterate through the connections states to determine the maximum `delayPeriod` for the channel path and verify that the counterparty consensus state on the receiving chain satisfies the delay requirement.
+4. Iterate through connection state proofs and verify each connectionEnd is in the OPEN state and check that the connection ids match the channel connectionHops.
+5. Verify the intermediate state proofs. Starting with known `ConsensusState[0]` at the given `proofHeight` on `Chain[1]` prove the prior chain's consensus and connection state.
+6. Verify that the client id in each consensus state proof key matches the client id in the ConnectionEnd in the previous connection state proof.
+7. Repeat step 5, proving `ConsensusState[i]`, and `Conn[i,i-1]` where `i` is the proof index starting with the consensus state on `Chain[2]`. `ConsensusState[1]` is already known on `Chain[0]`. Note that chains are indexed from executing (verifying) chain to  and proofs are indexed in the opposite direction to match the connectionHops ordering.
+   - Verify ParseClientID(ConsensusProofs[i].Key) == ConnectionEnd.ClientID
+   - ConsensusProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConsensusProofs[i].Key, ConsensusProofs[i].Value)
+   - ConnectionProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConnectionProofs[i].Key, ConnectionProofs[i].Value)
+   - ConsensusState = ConsensusProofs[i].Value
+   - i++
```

Comment:
these could be done in one pass, no?

Reply:
github missed the line selection, i am referring to 4 and 5-7.

Reply:
on the other hand, thinking about it more, maybe it is better to have it this way as we catch in a cheap way bad connection on last hops without first verifying proofs for nothing.

Reply:
yeah, step 4 iterates over only the connection state proofs (without verifying). 5-7 iterates over consensus states and uses the consensus states to recursively verify consensus and connection proofs.